src/core/socket.c

```diff
@@ -2039,13 +2089,29 @@ static void socket_enter_running(Socket *s, int cfd) {
         } else {
                 _cleanup_free_ char *prefix = NULL, *instance = NULL, *name = NULL;
                 Service *service;
+                SocketPeer *p;
 
                 if (s->n_connections >= s->max_connections) {
                         log_unit_warning(UNIT(s), "Too many incoming connections (%u), refusing connection attempt.", s->n_connections);
                         safe_close(cfd);
                         return;
                 }
 
+                if (s->max_connections_per_source > 0) {
+
+                        r = socket_find_peer(s, cfd, &p);
+                        if (r < 0) {
+                                safe_close(cfd);
+                                return;
+                        }
+
+                        if (p->n_ref > s->max_connections_per_source) {
+                                log_unit_warning(UNIT(s), "Too many incoming connections (%u) from source, refusing connection attempt.", p->n_ref);
+                                safe_close(cfd);
+                                return;
+                        }
+                }
```

Comment:
memory management appears borked here. You need to make sure that the socket peer object is automatically released when something fails here, if it was just created for this. My recommendation would be to make sure that socket_find_peer() always increases the ref cntr of the returned object by one on success, and the code here uses `_cleanup_(socket_peer_unrefp)` when declaring the SocketPeer, to ensure it is released again if something fails here...